contracts/poap/src/contract.rs

```diff
@@ -0,0 +1,1086 @@
+        assert_eq!(
+            ContractError::EndTimeBeforeCurrentTime {
+                current_time: env.block.time,
+                end_time: Timestamp::from_seconds(end)
+            },
+            init_result.unwrap_err()
+        );
+    }
+
+    #[test]
+    fn instantiate_with_event_end_equal_current_time() {
+        let mut deps = mock_dependencies();
+        let env = mock_env();
+        let info = mock_info(ADMIN, &vec![]);
+        let mut init_msg = get_valid_init_msg(1);
+
+        let start = env.block.time.seconds() - 200;
+        let end = env.block.time.nanos();
+        init_msg.event_info.start_time = Timestamp::from_seconds(start);
+        init_msg.event_info.end_time = Timestamp::from_nanos(end);
+
+        let init_result = instantiate(deps.as_mut(), env, info, init_msg);
+        assert_eq!(
+            ContractError::EndTimeBeforeCurrentTime {
+                current_time: Timestamp::from_nanos(end),
+                end_time: Timestamp::from_nanos(end),
+            },
+            init_result.unwrap_err()
+        );
+    }
+
+    #[test]
+    fn instantiate_with_event_start_after_end() {
+        let mut deps = mock_dependencies();
+        let env = mock_env();
+        let info = mock_info(ADMIN, &vec![]);
+        let mut init_msg = get_valid_init_msg(1);
+
+        // Create a start time 200 seconds after the current block time
+        let start = env.block.time.seconds() + 200;
+        let end = env.block.time.seconds() + 100;
+        init_msg.event_info.start_time = Timestamp::from_seconds(start);
+        // Start time 100 seconds before the event start time
+        init_msg.event_info.end_time = Timestamp::from_seconds(end);
+
+        let init_result = instantiate(deps.as_mut(), env, info, init_msg);
+        assert_eq!(
+            ContractError::StartTimeAfterEndTime {
+                start: Timestamp::from_seconds(start),
+                end: Timestamp::from_seconds(end),
+            },
+            init_result.unwrap_err()
+        );
+    }
+
+    #[test]
+    fn instantiate_with_event_start_eq_end() {
+        let mut deps = mock_dependencies();
+        let env = mock_env();
+        let info = mock_info(ADMIN, &vec![]);
+        let mut init_msg = get_valid_init_msg(1);
+
+        // Create a start time 200 seconds after the current block time
+        let start = env.block.time.seconds() + 200;
+        init_msg.event_info.start_time = Timestamp::from_seconds(start);
+        init_msg.event_info.end_time = Timestamp::from_seconds(start);
+
+        let init_result = instantiate(deps.as_mut(), env, info, init_msg);
+        assert_eq!(
+            ContractError::StartTimeAfterEndTime {
+                start: Timestamp::from_seconds(start),
+                end: Timestamp::from_seconds(start),
+            },
+            init_result.unwrap_err()
+        );
+    }
+
+    #[test]
+    fn instantiate_with_invalid_poap_uri() {
```

Comment:
as above

Reply:
I can write the tests that ensure that the `validate` method of `InstantiateMsg` and `ExecuteMsg` is implemented correctly but I do not understend how you wanna remove the tests.

Reply:
As you said, these test can be implemented with `validate` in `msg.rs`, so I think they can be moved to there. As a result, in `contract.rs` only need to have a test called `instantiate_with_invalid_msg_error`.